omnibus/resources/agent/msi/cal/CustomAction.cpp

```diff
@@ -0,0 +1,285 @@
+
+
+int CreateUser(std::wstring& name, std::wstring& comment, bool writePassToReg = false);
+DWORD DeleteUser(std::wstring& name);
+DWORD DeleteSecretsRegKey();
+
+#ifdef CA_CMD_TEST
+#define LOGMSG_STANDARD 0
+void WcaLog(int type, const char * fmt...)
+{
+	va_list args;
+	va_start(args, fmt);
+	vprintf(fmt, args);
+	va_end(args);
+	printf("\n");
+}
+#else
+extern "C" UINT __stdcall AddDatadogSecretUser(
+	MSIHANDLE hInstall
+	)
+{
+	HRESULT hr = S_OK;
+	UINT er = ERROR_SUCCESS;
+
+	// that's helpful.  WcaInitialize Log header silently limited to 32 chars 
+	hr = WcaInitialize(hInstall, "CA: AddDatadogSecretUser");
+	ExitOnFailure(hr, "Failed to initialize");
+
+	WcaLog(LOGMSG_STANDARD, "Initialized.");
+
+	er = CreateUser(secretUserUsername, secretUserDescription, true);
+	if (0 != er) {
+		hr = -1;
+	}
+
+LExit:
+	er = SUCCEEDED(hr) ? ERROR_SUCCESS : ERROR_INSTALL_FAILURE;
+	return WcaFinalize(er);
+}
+
+extern "C" UINT __stdcall RemoveDatadogSecretUser(MSIHANDLE hInstall) {
+	HRESULT hr = S_OK;
+	UINT er = ERROR_SUCCESS;
+
+	// that's helpful.  WcaInitialize Log header silently limited to 32 chars
+	hr = WcaInitialize(hInstall, "CA: RemoveDatadogSecretUser");
+	ExitOnFailure(hr, "Failed to initialize");
+
+	WcaLog(LOGMSG_STANDARD, "Initialized.");
+
+	er = DeleteUser(secretUserUsername);
+	if (0 != er) {
+		hr = -1;
+		goto LExit;
+	} 
+	er = DeleteSecretsRegKey();
+	if (0 != er) {
+		hr = -1;
+		goto LExit;
+	}
+
+
+LExit:
+	er = SUCCEEDED(hr) ? ERROR_SUCCESS : ERROR_INSTALL_FAILURE;
+	return WcaFinalize(er);
+
+}
+
+
+
+// DllMain - Initialize and cleanup WiX custom action utils.
+extern "C" BOOL WINAPI DllMain(
+	__in HINSTANCE hInst,
+	__in ULONG ulReason,
+	__in LPVOID
+	)
+{
+	switch(ulReason)
+	{
+	case DLL_PROCESS_ATTACH:
+		WcaGlobalInitialize(hInst);
+		// initialize random number generator
+		srand(GetTickCount());
+		break;
+
+	case DLL_PROCESS_DETACH:
+		WcaGlobalFinalize();
+		break;
+	}
+
+	return TRUE;
+}
+#endif // CA_CMD_TEST
+#define MIN_PASS_LEN 12
+#define MAX_PASS_LEN 18
+
+wchar_t * GeneratePassword() {
+	static wchar_t passbuf[MAX_PASS_LEN + 1];
+	
+	const wchar_t * availLower = L"abcdefghijklmnopqrstuvwxyz";
+	const wchar_t * availUpper = L"ABCDEFGHIJKLMNOPQRSTUVWXYZ"; 
+	const wchar_t * availNum = L"1234567890";
+	const wchar_t * availSpec = L"()`~!@#$%^&*-+=|{}[]:;'<>,.?/";
+
+#define CHARTYPE_LOWER 0
+#define CHARTYPE_UPPER 1
+#define CHARTYPE_NUMBER 2
+#define CHARTYPE_SPECIAL 3
+	const wchar_t * classes[] = {
+		availLower,
+		availUpper,
+		availNum,
+		availSpec,
+	};
+	size_t classlengths[] = {
+		wcslen(availLower),
+		wcslen(availUpper),
+		wcslen(availNum),
+		wcslen(availSpec)
+	};
+	int numtypes = sizeof(classes) / sizeof(wchar_t*);
+
+	int usedClasses[] = { 0, 0, 0, 0 };
+
+	// we'll do a random length password between 12 and 18 chars
+	int len = (rand() % (MAX_PASS_LEN - MIN_PASS_LEN)) + MIN_PASS_LEN;
+	int times = 0;
+	do {
```

Comment:
I don't think you want random_shuffle; b/c that precludes characters being reused.  We want it to be truly random.

The change to have more letters than num/chars was just preference, because some of the passwords were coming in really unreadable.

The reason we don't want to take ~30 chars from one string is that Windows has password complexity requirements by default.  If you just take ~30 chars you may not meet the complexity requirements.  Sorting out lower, upper, number, & special chars and making sure there are >= 2 of each makes sure we meet complexity requirements. As I said, the (artificial) requirement of more letters than nums/spec chars was just visual for me.

Reply:
I didn't knew about the windows requirements.

For random_shuffle, my idea was to add to a string in order between 2 and 4 lowercase characters, then upper case, then number and finally special ones. And at the end to random_shuffle that string. This way, we generate a password only once, that match windows requirements. What do you think ?

Reply:
I'm not sure how your suggestion differs greatly from mine (other than the fact that it sometimes takes >1 "tries" to get a password that meets complexity requirements).